### ISSUE TITLE
JFrog CLI 1.39 deletes sphinx CSS subdirectories etc. during uploads with rt upload

### ISSUE THREAD

**Issue by tmwong2003** [jfrog/jfrog-cli#808]
**Describe the bug**
We are observing an issue with uploading documentation created with `sphinx`. Our generated documentation directory looks like this:

```
$ ls -al
total 80
drwxr-xr-x  11 theodorew  1916    352 Aug 20 03:11 .
drwxr-xr-x   4 theodorew  1916    128 Aug 20 01:53 ..
-rw-r--r--   1 theodorew  1916    230 Aug 20 02:01 .buildinfo
drwxr-xr-x   4 theodorew  1916    128 Aug 20 03:12 .doctrees
drwxr-xr-x   3 theodorew  1916     96 Aug 20 00:58 _sources
drwxr-xr-x  18 theodorew  1916    576 Aug 20 00:58 _static
-rw-r--r--   1 theodorew  1916   4352 Aug 20 02:01 genindex.html
-rw-r--r--   1 theodorew  1916  10105 Aug 20 02:01 index.html
-rw-r--r--   1 theodorew  1916    327 Aug 20 02:01 objects.inv
-rw-r--r--   1 theodorew  1916   4145 Aug 20 02:01 search.html
-rw-r--r--   1 theodorew  1916   1594 Aug 20 02:01 searchindex.js
```

Invoking the JFrog CLI v1.38.4 yields a properly populated documentation directory on our local Artifactory server, but invoking JFrog CLI v1.39 in the same way drops the subdirectories after upload, include the `_static` subdirectory containing CSS files, etc.

**To Reproduce**
Invoke JFrog CLI from the `jfrog/jfrog-cli-go` Docker image:
```
jfrog rt config --interactive=false --url https://artifactory.local.example.net/artifactory --user $${ARTIFACTORY_USERNAME} --password $${ARTIFACTORY_PASSWORD}
jfrog rt upload ./ example-docs-local/package/latest/ --flat="false" --quiet="true" --sync-deletes="example-docs-local/package/latest/"
```

**Expected behavior**
We should see the following, as produced by v1.38.4:
```
Index of example-docs-local/pacakge/2020.224.0-2-ga9d350d

Name            Last modified      Size
../
.doctrees/       20-Aug-2020 19:49    -
_sources/        20-Aug-2020 19:49    -
_static/         20-Aug-2020 19:49    -
.buildinfo       20-Aug-2020 19:49  230 bytes
genindex.html    20-Aug-2020 19:49  4.25 KB
index.html       20-Aug-2020 19:49  9.87 KB
objects.inv      20-Aug-2020 19:49  327 bytes
search.html      20-Aug-2020 19:49  4.05 KB
searchindex.js   20-Aug-2020 19:49  1.56 KB
```
What actually see is the following, as produced by v1.39:
```
Index of example-docs-local/package/2020.224.0-1-g14a5e0b

Name            Last modified      Size
../
.buildinfo       20-Aug-2020 10:35  230 bytes
genindex.html    20-Aug-2020 10:35  4.25 KB
index.html       20-Aug-2020 10:35  9.87 KB
objects.inv      20-Aug-2020 10:35  327 bytes
search.html      20-Aug-2020 10:35  4.05 KB
searchindex.js   20-Aug-2020 10:35  1.56 KB
```

**Screenshots**
The log from v1.39 shows that the `rt upload` operations deletes all of the subdirectories, but gives no explanation as to why:
```
[... Lots of upload lines ...]
[Info] [Thread 2] Uploading artifact: .doctrees/environment.pickle
204	[Info] [Thread 0] Uploading artifact: .buildinfo
205	[Info] Searching artifacts...
206	[Info] Found 3 artifacts.
207	[Info] [Thread 1] Deleting example-docs-local/package/latest/_static/
208	[Info] [Thread 0] Deleting example-docs-local/package/latest/_sources/
209	[Info] [Thread 2] Deleting example-docs-local/package/latest/.doctrees/
210	{
211	  "status": "success",
212	  "totals": {
213	    "success": 92,
214	    "failure": 0
215	  }
216	}
```

**Versions**
- JFrog CLI version: 1.39
- JFrog CLI operating system: `jfrog/jfrog-cli-go:1.39`
- Artifactory Version: 7.5.5 rev 70505900

**Comment by Or-Gabay** [comment]
@tmwong2003, I couldn’t reproduce this issue as I followed your steps. It will help if you run the following steps:
1. Run `jfrog -v` to make sure you’re using version 1.39.0.
1. Create a new empty directory and cd into it.
2. Run `jfrog rt dl “tmwong2003/package/latest/(*)” {1} --url=https://ecosysjfrog.jfrog.io/artifactory --user=tmwong2003 --password=tmwong2003 --flat=true`
Now you should see the following directory structure locally:
```
.
├── 1
├── 2
├── 3
└── _static
    ├── 1
    └── details1
```
3. rename the file `details1` -> `details2`
4. Run `jfrog rt upload ./ tmwong2003/package/latest/ --flat=“false” --quiet=“true” --sync-deletes=“tmwong2003/package/latest/” --url=https://ecosysjfrog.jfrog.io/artifactory --user=tmwong2003 --password=tmwong2003`

If you manage to see the `_static` folder in Artifactory, with the updated file `details2` then it’s probably a bug related to the `--sync-deletes` functionality, that is revealed for your unique files structures. It will be helpful if you can share one of your folders to allow me to reproduce this issue.
You can zip the directory structure, and attach it to this GitHub issue. Please also try to provide precise steps to reproduce the issue.
I’m looking forward to hearing from you.

**Comment by milas** [comment]
I'm also seeing similar issues with JFrog CLI 1.39.1.

The issue appears to be with subfolders? My upload is something like:
```
docs/
  latest/
    root_file.txt
    subfolder_1/
      sub_1.txt
    subfolder_2/
      sub_2.txt
    subfolder_3/
      sub_3.txt
```
Only `root_file.txt` survives - previously, all the content in subfolders would also remain as expected.

My command is:
```
jfrog rt u --recursive --flat=false --sync-deletes="repo/docs/latest" "*" "repo/docs/latest"
```

When running with `JFROG_CLI_LOG_LEVEL=DEBUG`, here's what I see happening:
1. Lots of uploads (HTTP PUT) with `;sync.deletes.timestamp=1598458115311` -- all artifacts get uploaded successfully
2. Searches for files that are NOT for the timestamp:
     ```
     [Debug] Searching Artifactory using AQL query:
     items.find({"@sync.deletes.timestamp":{"$ne":"1598458115311"},"type":"any","path":{"$ne":"."},"$or":[{"$and":[{"repo":"docs","path":"docs/latest","name":{"$match":"*"}}]},{"$and":[{"repo":"repo","path":{"$match":"docs/latest/*"},"name":{"$match":"*"}}]}]}).include("name","repo","path","actual_md5","actual_sha1","size","type","modified","created")
    ```
3. Searches for files that ARE for the tiemstamp:
    ```
    [Debug] Searching Artifactory using AQL query:
    items.find({"@sync.deletes.timestamp":"1598458115311","type":"any","path":{"$ne":"."},"$or":[{"$and":[{"repo":"repo","path":"docs/latest","name":{"$match":"*"}}]},{"$and":[{"repo":"repo","path":{"$match":"docs/latest/*"},"name":{"$match":"*"}}]}]}).include("name","repo","path","actual_md5","actual_sha1","size","type","modified","created").sort({"$asc":["repo","path","name"]})
    ```
4. Finds 3 "artifacts" which appear to be the 3 subfolders which are then deleted!

**Comment by Or-Gabay** [comment]
@milas, It will be helpful if you can share one of your folders to allow me to reproduce this issue.
You can zip the directory structure, and attach it to this GitHub issue.

**Comment by tmwong2003** [comment]
> Run jfrog -v to make sure you’re using version 1.39.0.
bash-5.0# jfrog -v
jfrog version 1.39.0
> Create a new empty directory and cd into it.
bash-5.0# mkdir test; cd test
> Run jfrog rt dl “tmwong2003/package/latest/(*)” {1} --url=https://ecosysjfrog.jfrog.io/artifactory --user=tmwong2003 --password=tmwong2003 --flat=true
> Now you should see the following directory structure locally:
> .
> ├── 1
> ├── 2
> ├── 3
> └── _static
>     ├── 1
>     └── details1
bash-5.0# jfrog rt dl "tmwong2003/package/latest/(*)" {1} --url=https://ecosysjfrog.jfrog.io/artifactory --user=tmwong2003 --password=tmwong2003 --flat=true
[Info] Searching items to download...
[Info] [Thread 2] Downloading tmwong2003/package/latest/2
[Info] [Thread 1] Downloading tmwong2003/package/latest/3
[Info] [Thread 0] Downloading tmwong2003/package/latest/_static/1
[Info] [Thread 0] Downloading tmwong2003/package/latest/1
[Info] [Thread 1] Downloading tmwong2003/package/latest/_static/details1
{
  "status": "success",
  "totals": {
    "success": 5,
    "failure": 0
  }
}
> rename the file details1 -> details2
bash-5.0# mv _static/details1 _static/details2
> Run jfrog rt upload ./ tmwong2003/package/latest/ --flat=“false” --quiet=“true” --sync-deletes=“tmwong2003/package/latest/” --url=https://ecosysjfrog.jfrog.io/artifactory --user=tmwong2003 --password=tmwong2003
bash-5.0# jfrog rt upload ./ tmwong2003/package/latest/ --flat="false" --quiet="true" --sync-deletes="tmwong2003/package/latest/" --url=https://ecosysjfrog.jfrog.io/artifactory --user=tmwong2003 --password=tmwong2003
[Info] [Thread 1] Uploading artifact: 3
[Info] [Thread 2] Uploading artifact: _static/details2
[Info] [Thread 0] Uploading artifact: _static/1
[Info] [Thread 1] Uploading artifact: 2
[Info] [Thread 2] Uploading artifact: 1
[Info] Searching artifacts...
[Info] Found 1 artifact.
[Info] [Thread 2] Deleting tmwong2003/package/latest/_static/details1
{
  "status": "success",
  "totals": {
    "success": 5,
    "failure": 0
  }
}
I see that https://ecosysjfrog.jfrog.io/artifactory/tmwong2003/package/latest/_static <https://ecosysjfrog.jfrog.io/artifactory/tmwong2003/package/latest/_static> exists; I’m clearing the contents of the docs with management to make sure I can send the tarball.

**Comment by tmwong2003** [comment]
Trying to upload the problematic directory to the jfrog.io had the same problem as our local installation:
```
bash-5.0# jfrog -v
jfrog version 1.39.0
bash-5.0# ls -lF
total 28
drwxr-xr-x    3 root     root            96 Aug 28 02:39 _sources/
drwxr-xr-x   18 root     root           576 Aug 28 02:39 _static/
-rw-r--r--    1 root     root          3706 Aug 28 02:39 genindex.html
-rw-r--r--    1 root     root          5636 Aug 28 02:39 index.html
-rw-r--r--    1 root     root           286 Aug 28 02:39 objects.inv
-rw-r--r--    1 root     root          4121 Aug 28 02:39 search.html
-rw-r--r--    1 root     root           852 Aug 28 02:39 searchindex.js
bash-5.0# jfrog rt upload ./ tmwong2003/regression/latest/ --flat="false" --quiet="true" --sync-deletes="tmwong2003/regression/latest/" --url=https://ecosysjfrog.jfrog.io/artifactory --user=tmwong2003 --password=tmwong2003
[Info] [Thread 2] Uploading artifact: searchindex.js
[Info] [Thread 0] Uploading artifact: search.html
[Info] [Thread 1] Uploading artifact: objects.inv
[Info] [Thread 0] Uploading artifact: index.html
[Info] [Thread 1] Uploading artifact: genindex.html
[Info] [Thread 2] Uploading artifact: _static/underscore.js
[ ... bunch of files in _static ...]
Info] [Thread 0] Uploading artifact: _sources/index.rst.txt
[Info] [Thread 1] Uploading artifact: .doctrees/index.doctree
[Info] [Thread 0] Uploading artifact: .doctrees/environment.pickle
[Info] [Thread 2] Uploading artifact: .buildinfo
[Info] Searching artifacts...
[Info] Found 3 artifacts.
[Info] [Thread 2] Deleting tmwong2003/regression/latest/.doctrees/
[Info] [Thread 1] Deleting tmwong2003/regression/latest/_static/
[Info] [Thread 0] Deleting tmwong2003/regression/latest/_sources/
{
  "status": "success",
  "totals": {
    "success": 92,
    "failure": 0
  }
}
bash-5.0# 
```
I pushed a tarball [jfrog-test.tgz](https://ecosysjfrog.jfrog.io/artifactory/tmwong2003/jfrog-test.tgz) of the 
and attached the full log file [jfrog-upload.log](https://github.com/jfrog/jfrog-cli/files/5139750/jfrog-upload.log) of the upload session.

**Comment by milas** [comment]
@Or-Gabay My structure/setup is _very_ similar to that of @tmwong2003 -- if the example they shared is insufficient to repro, let me know and I can try to create a repro case as well, but it will be almost identical 🙂

**Comment by Or-Gabay** [comment]
First of all, thank you guys **both** for your collaboration, I have managed to fix the bug and we will release it soon.
Moreover, I have added a [JFrog-CLI]() binary (Linux edition) which includes the fix, feel free to try it out and share your results.

**Comment by eyalbe4** [comment]
@milas and @tmwong2003,
Version 1.39.2 is released and includes @Or-Gabay's fix for this issue.
Thanks again for helping us pinpoint and fix the issue. We're looking forward to your feedback for this patch release.
And of course, thanks @Or-Gabay for fixing the issue!

**Comment by tmwong2003** [comment]
@eyalbe4 @Or-Gabay Thank you for releasing the fix! I can confirm that JFrog 1.39.2 fixes the issue we saw.
